src/lib/swapper/swappers/ThorchainSwapper/utils/longTailHelpers.ts

```diff
@@ -53,3 +58,117 @@ export function getTradeType(
       return undefined
   }
 }
+
+// TODO: we need to fetch these contracts dynamically, as the whitelist can change
+export enum AggregatorContract {
+  TSAggregatorUniswapV3_100 = '0xbd68cbe6c247e2c3a0e36b8f0e24964914f26ee8',
+  TSAggregatorUniswapV3_500 = '0xe4ddca21881bac219af7f217703db0475d2a9f02',
+  TSAggregatorUniswapV3_3000 = '0x11733abf0cdb43298f7e949c930188451a9a9ef2',
+  TSAggregatorUniswapV3_10000 = '0xb33874810e5395eb49d8bd7e912631db115d5a03',
+}
+
+export const contractToFeeAmountMap: Record<AggregatorContract, FeeAmount> = {
+  [AggregatorContract.TSAggregatorUniswapV3_100]: FeeAmount.LOWEST,
+  [AggregatorContract.TSAggregatorUniswapV3_500]: FeeAmount.LOW,
+  [AggregatorContract.TSAggregatorUniswapV3_3000]: FeeAmount.MEDIUM,
+  [AggregatorContract.TSAggregatorUniswapV3_10000]: FeeAmount.HIGH,
+}
+
+export const feeAmountToContractMap: Record<FeeAmount, AggregatorContract> = Object.entries(
+  contractToFeeAmountMap,
+).reduce(
+  (acc, [key, value]) => ({ ...acc, [value]: key }),
+  {} as Record<FeeAmount, AggregatorContract>,
+)
+
+export const generateV3PoolAddressesAcrossFeeRange = (
+  factoryAddress: string,
+  tokenA: Token,
+  tokenB: Token,
+): Map<Address, FeeAmount> => {
+  const poolAddresses = new Map<Address, FeeAmount>()
+  Object.values(FeeAmount)
+    .filter((value): value is FeeAmount => typeof value === 'number')
+    .forEach((fee: FeeAmount) => {
+      const poolAddress = computePoolAddress({
+        factoryAddress,
+        tokenA,
+        tokenB,
+        fee,
+      })
+      poolAddresses.set(poolAddress as Address, fee as FeeAmount)
+    })
+  return poolAddresses
+}
+
+type ContractData = {
+  fee: FeeAmount
+  tokenIn: Address
+  tokenOut: Address
+}
+
+export const getContractDataByPool = async (
+  poolAddresses: Map<Address, FeeAmount>,
+  publicClient: PublicClient,
+  tokenAAddress: string,
+  tokenBAddress: string,
+): Promise<Map<Address, ContractData>> => {
+  const poolContracts = new Map<Address, ContractData>()
+  await Promise.all(
+    Array.from(poolAddresses.entries()).map(async ([address, feeAmount]) => {
+      const poolContract = getContract({
+        abi: IUniswapV3PoolABI,
+        address,
+        publicClient,
+      })
+      try {
+        const [token0, token1]: [Address, Address] = await Promise.all([
+          poolContract.read.token0(),
+          poolContract.read.token1(),
+        ])
```

Comment:
if we stored the tokens in the result of `generateV3PoolAddressesAcrossFeeRange` such that the index generated is 
```
Map<Address, { tokenA: Address; tokenB: Address; fee: FeeAmount }>
```

.. then we can avoid 2 http requests here

Reply:
oh i see, this is actually required to determine the order of the tokens. Ignore me

Reply:
not a big deal, but all tokens in Uniswap should be sorted alphabetically, so we could avoid the calls. 

see here:
https://github.com/Uniswap/v3-core/blob/d8b1c635c275d2a9450bd6a78f3fa2484fef73eb/contracts/UniswapV3Factory.sol#L41

Reply:
Addressed in https://github.com/shapeshift/web/pull/5902 🙏